camel/toolkits/function_tool.py

```diff
@@ -151,23 +237,49 @@ class FunctionTool:
     provide a user-defined tool schema to override.
 
     Args:
-        func (Callable): The function to call.The tool schema is parsed from
-            the signature and docstring by default.
-        openai_tool_schema (Optional[Dict[str, Any]], optional): A user-defined
-            openai tool schema to override the default result.
+        func (Callable): The function to call. The tool schema is parsed from
+            the function signature and docstring by default.
+        openai_tool_schema (Optional[Dict[str, Any]], optional): A
+            user-defined OpenAI tool schema to override the default result.
             (default: :obj:`None`)
+        use_schema_assistant (Optional[bool], optional): Whether to enable the
+            use of a schema assistant model to automatically generate the
+            schema if validation fails or no valid schema is provided.
+            (default: :obj:`False`)
+        schema_assistant_model (Optional[BaseModelBackend], optional): An
+            assistant model (e.g., an LLM model) used to generate the schema
+            if `use_schema_assistant` is enabled and no valid schema is
+            provided. (default: :obj:`None`)
+        schema_generation_max_retries (int, optional): The maximum
+            number of attempts to retry schema generation using the schema
+            assistant model if the previous attempts fail. (default: 2)
     """
 
     def __init__(
         self,
         func: Callable,
         openai_tool_schema: Optional[Dict[str, Any]] = None,
+        use_schema_assistant: Optional[bool] = False,
+        schema_assistant_model: Optional[BaseModelBackend] = None,
+        schema_generation_max_retries: int = 2,
     ) -> None:
         self.func = func
         self.openai_tool_schema = openai_tool_schema or get_openai_tool_schema(
             func
         )
 
+        if use_schema_assistant:
+            schema = self.generate_openai_tool_schema(
+                schema_generation_max_retries, schema_assistant_model
+            )
+            if schema:
+                self.openai_tool_schema = schema
+            else:
+                raise ValueError(
+                    f"Failed to generate valid schema for "
+                    f"{self.func.__name__}."
+                )
```

Comment:
In the previous [discussion](https://github.com/camel-ai/camel/pull/1070#discussion_r1810476373), Guohao suggested that the `validate_openai_tool_schema ` could be omitted so that even if the original schema has issues, the generation will fix it.

Reply:
Got it, then I think we need to give some warming msg to user, let them know once `use_schema_assistant ` is `True`, the tool schema would be overwritten by agent